### PR TITLE
Using the dag implementation in pfs internal api server

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

File: etc/kube/pachyderm.json [hunk 1]
```diff
@@ -325,7 +323,7 @@
                 "value": "/pach"
               },
               {
-                "name": "PFS_NUM_SHARDS",
```
Comment (jdoliner) [inline etc/kube/pachyderm.json]: This is the actual change, I'd changed the flag pachd looks for in pachd but not in the manifest. This fixes the bug.

File: etc/kube/pachyderm.json [hunk 2]
```diff
@@ -348,24 +346,7 @@
   },
   "status": {
     "replicas": 0
-  }
```
Comment (jdoliner) [inline etc/kube/pachyderm.json]: Change due to pachctl being updated in a prior commit but us not running `make kube-cluster-assets`.

File: etc/kube/pachyderm.json [hunk 3]
```diff
@@ -1,6 +1,4 @@
 {
```
Comment (jdoliner) [inline etc/kube/pachyderm.json]: Most of the changes to this file are superficial due to json encodings not being deterministic.

File: src/pfs/drive/driver.go [hunk 4]
```diff
@@ -17,10 +18,9 @@ type driver struct {
 	blockAddress    string
 	blockClient     pfs.BlockAPIClient
 	blockClientOnce sync.Once
-	started         diffMap
-	finished        diffMap
-	leaves          diffMap // commits with no children
-	branches        diffMap
```
Comment (jdoliner) [inline src/pfs/drive/driver.go]: These 4 lines are really nice to get refactored. They were quite kludgy.

File: src/pfs/drive/driver.go [hunk 5]
```diff
@@ -17,10 +18,9 @@ type driver struct {
...
+	diffs           diffMap
```
Comment (jdoliner) [inline src/pfs/drive/driver.go]: We now have a single diffMap, much saner.

File: src/pfs/drive/driver.go [hunk 6]
```diff
@@ -17,10 +18,9 @@ type driver struct {
...
+	dags            map[string]*dag.DAG
+	branches        map[string]map[string]string
```
Comment (jdoliner) [inline src/pfs/drive/driver.go]: These fields store indexes in to the diffs structure.

File: src/pfs/drive/driver.go [hunk 7]
```diff
@@ -684,6 +639,48 @@ func (d *driver) lastRef(file *pfs.File, shard uint64) *pfs.Commit {
...
+// canonicalCommit finds the canonical way of referring to a commit
+func (d *driver) canonicalCommit(commit *pfs.Commit) (*pfs.Commit, error) {
+	if _, ok := d.branches[commit.Repo.Name]; !ok {
+		return nil, fmt.Errorf("repo %s not found", commit.Repo.Name)
+	}
+	if commitID, ok := d.branches[commit.Repo.Name][commit.Id]; ok {
+		return pfsutil.NewCommit(commit.Repo.Name, commitID), nil
+	}
+	return commit, nil
+}
```
Comment (jdoliner) [inline src/pfs/drive/driver.go]: canonicalCommit implements the branch resolving logic

File: src/pfs/drive/driver.go [hunk 8]
```diff
@@ -812,34 +805,3 @@ func (d diffMap) pop(diff *pfs.Diff) *pfs.DiffInfo {
...
-	d.branches[repo.Name] = make(map[uint64]map[string]*pfs.DiffInfo)
-}
-
-// canonicalCommit finds the canonical way of referring to a commit
-func (d *driver) canonicalCommit(commit *pfs.Commit, shard uint64) *pfs.Commit {
-	if diffInfo, ok := d.branches.get(&pfs.Diff{Commit: commit, Shard: shard}); ok {
-		return diffInfo.Diff.Commit
-	}
-	return commit
-}
```
Comment (jdoliner) [inline src/pfs/drive/driver.go]: 3 annoying functions that we no longer have to deal with :)

Comment (jdoliner) [inline src/pfs/drive/driver.go]: Although I guess some of these just moved around.

File: src/pfs/pfsutil/pfsutil.go [hunk 9]
```diff
@@ -39,6 +39,13 @@ func NewBlock(hash string) *pfs.Block {
 	}
 }
 
+func NewDiff(repoName string, commitID string, shard uint64) *pfs.Diff {
+	return &pfs.Diff{
+		Commit: NewCommit(repoName, commitID),
+		Shard:  shard,
+	}
+}
+
```
Comment (jdoliner) [inline src/pfs/pfsutil/pfsutil.go]: Convenience function I finally decided to add.
Actually I'm going to go through now and use this in some places.

File: src/pfs/server/server_test.go [hunk 10]
```diff
@@ -0,0 +1,165 @@
+package server
```
Comment (jdoliner) [inline src/pfs/server/server_test.go]: This is the good stuff right here. Standalone pfs tests :).

File: src/pkg/deploy/assets/assets.go [hunk 11]
```diff
@@ -93,7 +93,7 @@ func PachdRc(shards uint64, secrets bool) *api.ReplicationController {
 									Value: "/pach",
 								},
 								{
-									Name:  "PFS_NUM_SHARDS",
```
Comment (jdoliner) [inline src/pkg/deploy/assets/assets.go]: This was a latent bug.

File: src/pkg/require/require.go [hunk 12]
```diff
@@ -39,6 +39,12 @@ func NoError(tb testing.TB, err error, msgAndArgs ...interface{}) {
 	}
 }
 
+func ErrorExpected(tb testing.TB, err error, msgAndArgs ...interface{}) {
+	if err == nil {
+		fatal(tb, msgAndArgs, "Error is expected but got %v", err)
+	}
+}
+
```
Comment (jdoliner) [inline src/pkg/require/require.go]: Should this be called `YesError` to match `NoError`?

File: src/pfs/server/server_test.go [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+	pfsClient := getPfsClient(t)
+	repo := uniqueString("TestDisallowReadsDuringCommit")
+	require.NoError(t, pfsutil.CreateRepo(pfsClient, repo))
+	commit1, err := pfsutil.StartCommit(pfsClient, repo, "", "")
+	require.NoError(t, err)
+	_, err = pfsutil.PutFile(pfsClient, repo, commit1.Id, "foo", 0, strings.NewReader("foo\n"))
+	require.NoError(t, err)
+
+	// Make sure we can't get the file before the commit is finished
+	var readBuffer bytes.Buffer
```
Comment (jdoliner) [inline src/pfs/server/server_test.go]: This is a bit OCD but could we have this be the same buffer as below and just reinitialize?

**Comment by jdoliner** [comment]
Overview:

The big change is the addition of `src/pkg/dag` which implements some computations on directed acyclic graphs. This package is used by `src/pfs/driver` to implement pfs's version control semantics. Over time I expect `src/pkg/dag` to get more featureful and pfs will be able to lean on it harder. We'll also use it in pps to implement pipelines.

This pr also adds `src/pfs/server/server_test.go` which implements local pfs server tests. Being able to test the pfs layer locally should help us a lot. Their coverage isn't great right now but you gotta start somewhere.

File: etc/kube/pachyderm.json [hunk 14]
```diff
@@ -348,24 +346,7 @@
...
-      "app": "amazon-secret"
-    }
   },
-  "data": {
-    "bucket": "",
-    "id": "",
-    "secret": "",
-    "token": "",
-    "region": ""
-  }
```
Comment (jdoliner) [inline etc/kube/pachyderm.json]: This is a latent change from pachctl being updated without regenerating this file.

File: src/pkg/require/require.go (hunk 12 above)
Comment (sjezewski) [inline src/pkg/require/require.go]: Yup. Done and pushed.

**Pull request merged**